internal/beater/config/agentconfig.go

```diff
@@ -23,36 +23,103 @@ import (
 	"fmt"
 	"strings"
 	"time"
+
+	"github.com/pkg/errors"
+
+	"github.com/elastic/apm-server/internal/elasticsearch"
+	"github.com/elastic/elastic-agent-libs/config"
+	"github.com/elastic/elastic-agent-libs/logp"
+	"github.com/elastic/go-ucfg"
 )
 
-// KibanaAgentConfig holds remote agent config information
-type KibanaAgentConfig struct {
-	Cache Cache `config:"cache"`
+const msgInvalidConfigAgentCfg = "invalid value for `apm-server.agent.config.cache.expiration`, only accepting full seconds"
+
+// esCredentialsConfig is exposed to allow fine-tuned permission control
+// and is required when working with Fleet.
+type esCredentialsConfig struct {
+	Username string `config:"username"`
+	Password string `config:"password"`
+	APIKey   string `config:"api_key"`
+}
+
+func (c *esCredentialsConfig) Unpack(in **elasticsearch.Config) error {
```

Comment:
Using `**elasticsearch.Config` here so that user can call `.Unpack(&c.ESConfig)` as usual

Reply:
> The config that is passed into the Unpack should generally not be the one changed.

I think most of the `a.Unpack(&b)` usages mean "unpack what's inside a into b", meaning b is updated. e.g. https://github.com/elastic/apm-server/blob/main/internal/beater/config/sampling.go#L133
```
if err := outputESCfg.Unpack(&c.ESConfig); err != nil {
	return errors.Wrap(err, "error unpacking output.elasticsearch config for tail sampling")
}
```
It means that we apply outputESCfg to c.ESConfig. Therefore I'm not sure what's the problem with the `Unpack` implementation here.

> But only the username, password and api_key should be exposed to the users. Where are they exposed now? I also don't see any changes to the apm-server.yml anymore. The credentials still need to be exposed, otherwise they cannot be ingested via Fleet or the users themselves when using standalone Elastic Agent.

The username, password and api_key are still exposed at the same place but the changes to apm-server.yml have been removed because of Andrew's comment [here](https://github.com/elastic/apm-server/pull/9720#discussion_r1052760622). I am happy to add the changes for credentials in apm-server.yml back.

Reply:
The unpack issue is resolved by explicitly replacing the username, password and api_key fields when available.